Drivers/CAN/can.c

```diff
@@ -0,0 +1,289 @@
+                HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+                break;
+            case CAN_line_3:
+                /**CAN3 GPIO Configuration
+                PB4     ------> CAN3_RX
+                PB5     ------> CAN3_TX
+                */
+                GPIO_InitStruct.Pin = GPIO_PIN_4|GPIO_PIN_5;
+                GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+                GPIO_InitStruct.Pull = GPIO_NOPULL;
+                GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+                GPIO_InitStruct.Alternate = GPIO_AF9_CAN3;
+                HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+                break;
+            default:
+                /**CAN1 GPIO Configuration
+                PB8     ------> CAN1_RX
+                PB9     ------> CAN1_TX
+                */
+                GPIO_InitStruct.Pin = GPIO_PIN_8|GPIO_PIN_9;
+                GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
+                GPIO_InitStruct.Pull = GPIO_NOPULL;
+                GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
+                GPIO_InitStruct.Alternate = GPIO_AF9_CAN1;
+                HAL_GPIO_Init(GPIOB, &GPIO_InitStruct);
+                break;
+        }
+    }
+}
+
+
+void CAN_Handler_Init(CAN_HandleTypeDef* can_h, uint8_t line)
+{
+    switch(line)
+    {
+        case CAN_line_1:
+            can_h->Instance = CAN1;
+            break;
+        case CAN_line_2:
+            can_h->Instance = CAN2;
+            break;
+        case CAN_line_3:
+            can_h->Instance = CAN3;
+            break;
+        default:
+            can_h->Instance = CAN1;
+            break;
+    }
+    can_h->Init.Prescaler              = CAN_PRESCALER;
+    can_h->Init.Mode                   = CAN_MODE;
+    can_h->Init.SyncJumpWidth          = CAN_SYNC_JUMP_WIDTH;
+    can_h->Init.TimeSeg1               = CAN_TIME_SEG_1;
+    can_h->Init.TimeSeg2               = CAN_TIME_SEG_2;
+    can_h->Init.TimeTriggeredMode      = CAN_TIME_TRIGGERED_MODE;
+    can_h->Init.AutoBusOff             = CAN_AUTO_BUS_OFF;
+    can_h->Init.AutoWakeUp             = CAN_AUTO_WAKEUP;
+    can_h->Init.AutoRetransmission     = CAN_AUTO_RETRANSMISSION;
+    can_h->Init.ReceiveFifoLocked      = CAN_RECEIVE_FIFO_LOCKED;
+    can_h->Init.TransmitFifoPriority   = CAN_TRANSMIT_FIFO_PRIORITY;
+
+    CAN_Msp_Init(can_h, line);
+
+}
+
+
+uint8_t start_CAN()
+{
+    HAL_StatusTypeDef CAN_Status;
+
+    CAN_Handler_Init(CAN1, CAN_line_1);
+    CAN_Handler_Init(CAN2, CAN_line_2);
+    CAN_Handler_Init(CAN3, CAN_line_3);
+
+
+
+    CAN_Status = HAL_CAN_Start(CAN1);
+    if (CAN_Status!= HAL_OK)
+    {
+        return FAIL;
+    }
+    CAN_Status = HAL_CAN_Start(CAN2);
+    if (CAN_Status!= HAL_OK)
+    {
+        return FAIL;
+    }
+    CAN_Status = HAL_CAN_Start(CAN3);
+    if (CAN_Status!= HAL_OK)
+    {
+        return FAIL;
+    }
+    
+    return SUCCESS;
+}
+
+CAN_t CAN_setup(uint32_t id, uint8_t len, uint8_t* data, uint8_t bus, CAN_HandleTypeDef* can_line)
```

Comment:
This should return an error because it uses malloc, I'm pretty sure you can extract an error code out of malloc (i.e. fails if no memory and that should be trickled up)

Reply:
this is probably not a concern but what does Hal CAN init consist of. ie is it blocking, does it happen fast. Just get a wee bit concerned passing by refrence into a premade driver something that gets destroyed out of scope right after that function call